### PR TITLE
Fixes for policy traits and CUDA

### PR DESCRIPTION
**Pull request by davidbeckingsale**
@willkill07 @trws this is almost fixed, but the scan tests won't compile.

### CODE REVIEW

**Comment by willkill07** [comment]
@davidbeckingsale looking into it now

**Comment by davidbeckingsale** [comment]
Nice, thanks! It's breaking in the scan tests when it instantiates the storage stuff, but I can't work out why.

File: include/RAJA/internal/type_traits.hpp [hunk 1]
```diff
@@ -128,20 +128,20 @@ struct is_enum_same : std::false_type {
 template <typename T, T A>
 struct is_enum_same<T, A, A> : std::true_type {
 };
-template <typename Policy, Policy P>
+template <typename Policy, RAJA::Policy P>
 struct models_policy
     : std::integral_constant<bool,
-                             is_enum_same<Policy, Policy::policy, P>::value> {
+                             is_enum_same<RAJA::Policy, Policy::policy, P>::value> {
```
Comment (davidbeckingsale) [inline include/RAJA/internal/type_traits.hpp]: This line was the fix, I think. It seems like the template argument names were shadowing what you meant.

**Comment by davidbeckingsale** [comment]
You smushed your new numeric limits into this? We can merge my commit into that branch instead for cleanliness?

**Comment by willkill07** [comment]
I didn't mean to smudge it :( fixing now -- I resolved all other issues though (from my machine at least)

**Comment by davidbeckingsale** [comment]
Great, if you tidy this up I'll test on our machines and get this merged.

File: include/RAJA/internal/type_traits.hpp (hunk 1 above)
Comment (willkill07) [inline include/RAJA/internal/type_traits.hpp]: I ended up needing to rename it anyways -- the compiler didn't like when I was overly specific

**Comment by willkill07** [comment]
This got a little wonky (cherry-pick wasn't the catch all), but this should be good to go then

**Comment by willkill07** [comment]
GPU test output:

```
[wkillian@liz] ~/RAJA/build  (bugfix/cuda-policy-traits)
$ make test
Running tests...
Test project /home/wkillian/RAJA/build
      Start  1: layout-test.exe
 1/20 Test  #1: layout-test.exe ..................   Passed    0.00 sec
      Start  2: timer-test.exe
 2/20 Test  #2: timer-test.exe ...................   Passed    2.00 sec
      Start  3: multipolicy.exe
 3/20 Test  #3: multipolicy.exe ..................   Passed    0.00 sec
      Start  4: type-traits.exe
 4/20 Test  #4: type-traits.exe ..................   Passed    0.00 sec
      Start  5: CPUreduce-test.exe
 5/20 Test  #5: CPUreduce-test.exe ...............   Passed    0.53 sec
      Start  6: CPUnested-test.exe
 6/20 Test  #6: CPUnested-test.exe ...............   Passed    0.05 sec
      Start  7: CPUnested_reduce-test.exe
 7/20 Test  #7: CPUnested_reduce-test.exe ........   Passed    0.01 sec
      Start  8: indexset-tests.exe
 8/20 Test  #8: indexset-tests.exe ...............   Passed    0.49 sec
      Start  9: reduction-tests.exe
 9/20 Test  #9: reduction-tests.exe ..............   Passed    0.03 sec
      Start 10: scan-tests.exe
10/20 Test #10: scan-tests.exe ...................   Passed    0.03 sec
      Start 11: forall-tests.exe
11/20 Test #11: forall-tests.exe .................   Passed    0.51 sec
      Start 12: reducemin.exe
12/20 Test #12: reducemin.exe ....................   Passed    1.11 sec
      Start 13: reducemax.exe
13/20 Test #13: reducemax.exe ....................   Passed    1.08 sec
      Start 14: reducesum.exe
14/20 Test #14: reducesum.exe ....................   Passed    3.60 sec
      Start 15: traversal.exe
15/20 Test #15: traversal.exe ....................   Passed    0.45 sec
      Start 16: nested.exe
16/20 Test #16: nested.exe .......................   Passed    1.48 sec
      Start 17: reduceminloc.exe
17/20 Test #17: reduceminloc.exe .................   Passed    1.62 sec
      Start 18: reducemaxloc.exe
18/20 Test #18: reducemaxloc.exe .................   Passed    1.80 sec
      Start 19: forallN.exe
19/20 Test #19: forallN.exe ......................   Passed    0.90 sec
      Start 20: scan.exe
20/20 Test #20: scan.exe .........................   Passed    3.91 sec

100% tests passed, 0 tests failed out of 20

Total Test time (real) =  19.61 sec
```

File: include/RAJA/internal/type_traits.hpp [hunk 2]
```diff
@@ -117,54 +118,65 @@ struct is_wrapper_policy<Outer<Tags...>>
...
+                                    is_wrapper_policy<T>::value
+                                        || std::is_base_of<T,
+                                                           PolicyBase>::value> {
 };
 
 namespace detail
 {
 template <typename T, T A, T B>
-struct is_enum_same : std::false_type {
+struct is_enum_same : public std::false_type {
```
Comment (willkill07) [inline include/RAJA/internal/type_traits.hpp]: switched to verbosely defining inheritance levels

File: include/RAJA/internal/type_traits.hpp [hunk 3]
```diff
@@ -117,54 +118,65 @@ struct is_wrapper_policy<Outer<Tags...>>
...
-struct is_enum_same<T, A, A> : std::true_type {
+struct is_enum_same<T, A, A> : public std::true_type {
 };
-template <typename Policy, Policy P>
-struct models_policy
-    : std::integral_constant<bool,
-                             is_enum_same<Policy, Policy::policy, P>::value> {
+
+template <typename P_, Policy P>
+struct models_policy : public std::integral_constant<bool, is_enum_same<Policy, P_::policy, P>::value> {
```
Comment (willkill07) [inline include/RAJA/internal/type_traits.hpp]: Needed to fix names to avoid clobbering

File: test/include/data_storage.hxx [hunk 4]
```diff
@@ -4,13 +4,13 @@
...
 #include "RAJA/RAJA.hpp"
-#include "RAJA/util/defines.hpp"
 #include "RAJA/internal/type_traits.hpp"
+#include "RAJA/util/defines.hpp"
 
 namespace internal
 {
 
-template <typename ExecPolicy, typename T, bool gpu = false>
+template <typename ExecPolicy, typename T, bool gpu>
```
Comment (willkill07) [inline test/include/data_storage.hxx]: removed default argument and explicitly specified below

File: test/unit/test-type-traits.cxx [hunk 5]
```diff
@@ -0,0 +1,27 @@
+#include "RAJA/RAJA.hpp"
+
+#include "RAJA/internal/type_traits.hpp"
+
+#include "gtest/gtest.h"
+
+#ifdef RAJA_USE_OPENMP
```
Comment (willkill07) [inline test/unit/test-type-traits.cxx]: Fix test if no OpenMP or CUDA

**Comment by davidbeckingsale** [comment]
Thanks @willkill07 - I'll test this on Power and merge tomorrow.

**Comment by davidbeckingsale** [comment]
All looking good here, merging.

**Pull request merged**